kettle/stream.py

```diff
@@ -0,0 +1,298 @@
+    for ack_id, message in results:
+        if message.attributes['event_type'] != 'OBJECT_FINALIZE':
+            acks.append(ack_id)
+            continue
+        obj = message.attributes['object_id']
+        if not obj.endswith('/finished.json'):
+            acks.append(ack_id)
+            continue
+        job, build = obj[:-len('/finished.json')].rsplit('/', 1)
+        job = 'gs://%s/%s' % (message.attributes['bucket_id'], job)
+        todo.append((ack_id, job, build))
+
+    return acks, todo
+
+
+def get_started_finished(gcs_client, db, todo):
+    """Download started/finished.json from build dirs in todo."""
+    acks = []
+    build_dirs = []
+    pool = multiprocessing.pool.ThreadPool(16)
+    try:
+        for ack_id, (build_dir, started, finished) in pool.imap_unordered(
+                lambda (ack_id, job, build): (ack_id, gcs_client.get_started_finished(job, build)),
+                todo):
+            # build_dir, started, finished = gcs_client.get_started_finished(job, build)
+            if finished:
+                if not db.insert_build(build_dir, started, finished):
+                    print('already present??')
+                start = time.localtime(started.get('timestamp', 0) if started else 0)
+                print(build_dir, bool(started), bool(finished),
+                      time.strftime('%F %T %Z', start),
+                      finished and finished.get('result'))
+                build_dirs.append(build_dir)
+                acks.append(ack_id)
+            else:
+                print('???', build_dir, started, finished)
+    finally:
+        pool.close()
+    db.commit()
+    return acks, build_dirs
+
+
+def row_to_mapping(row, schema):
+    """Convert a dictionary to a list for bigquery.Table.insert_data.
+
+    Silly. See https://github.com/GoogleCloudPlatform/google-cloud-python/issues/3396
+    """
+    return [row.get(field.name, [] if field.mode == 'REPEATED' else None) for field in schema]
+
+
+def insert_data(table, rows_iter):
+    """Upload rows from rows_iter into bigquery table table.
+
+    rows_iter should return a series of (row_id, row dictionary) tuples.
+    The row dictionary must match the table's schema.
+
+    Returns the row_ids that were inserted.
+    """
+    emitted = set()
+
+    rows = []
+    row_ids = []
+
+    for row_id, row in rows_iter:
+        emitted.add(row_id)
+        if len(json.dumps(row)) > 1e6:
+            print('ERROR: row too long', row['path'])
+            continue
+        row = row_to_mapping(row, table.schema)
+        rows.append(row)
+        row_ids.append(row_id)
+
+    def insert(table, rows, row_ids):
+        """Insert rows with row_ids into table, retrying as necessary."""
+        while True:
+            try:
+                errors = table.insert_data(rows, row_ids, skip_invalid_rows=True)
+                break
+            except google.cloud.exceptions.ServerError:   # pylint: disable=no-member
```

Comment:
More work for you, but super useful for the next person who wants to help develop/maintain this since they only need to know how to install bazel, instead of a different cacophony of dependencies for each program.